src/sources/futures.rs

```diff
@@ -56,43 +116,187 @@ impl<T> Scheduler<T> {
+            //   runnables will be destroyed, preventing any unsoundness.
+            //
+            // - "If future is not 'static, borrowed variables must outlive its Runnable."
+            //
+            //   `F` is `'static`, so we don't have to worry about this one.
+            //
+            // - "If schedule is not Send and Sync, the task’s Waker must be used and dropped on the
+            //    original thread."
+            //
+            //   The schedule function uses a thread-safe MPSC channel to send the runnable back to the
+            //   origin thread. This means that the waker can be sent to another thread, which satisfies
+            //   this requirement.
+            //
+            // - "If schedule is not 'static, borrowed variables must outlive the task’s Waker."
+            //
+            //   `schedule` and the types it handles are `'static`, so we don't have to worry about
+            //   this one.
+            unsafe { builder.spawn_unchecked(move |_| future, schedule) }
+        };
+
+        // Insert the runnable into the set of active tasks.
+        active.insert(Active::Future(runnable.waker()));
+        drop(active_guard);
+
+        // Schedule the runnable and detach the task so it isn't cancellable.
+        runnable.schedule();
+        task.detach();
+
+        Ok(())
     }
 }
 
-/// Error generated when trying to schedule a future after the
-/// executor was destroyed.
-#[derive(thiserror::Error, Debug)]
-#[error("the executor was destroyed")]
-pub struct ExecutorDestroyed;
+impl Sender {
+    /// Send a runnable to the executor.
+    fn send(&self, runnable: Runnable<usize>) {
+        // Send on the channel.
+        if let Err(e) = self
+            .sender
+            .lock()
+            .unwrap_or_else(|e| e.into_inner())
+            .send(runnable)
+        {
+            // Make sure the runnable is never dropped.
+            std::mem::forget(e);
 
-#[derive(Debug)]
-struct ExecWaker {
-    ping: Ping,
+            // This shouldn't be able to happen since all of the tasks are destroyed before the
+            // executor is. This indicates a critical soundness bug.
+            std::process::abort();
+        }
+
+        // If the executor is already awake, don't bother waking it up again.
+        if self.notified.swap(true, Ordering::SeqCst) {
+            return;
+        }
+
+        // Wake the executor.
+        self.wake_up.ping();
+    }
 }
 
-impl ArcWake for ExecWaker {
-    fn wake_by_ref(arc_self: &Arc<ExecWaker>) {
-        arc_self.ping.ping();
+impl<T> Drop for Executor<T> {
+    fn drop(&mut self) {
+        let active = self.state.active.borrow_mut().take().unwrap();
+
+        // Wake all of the active tasks in order to destroy their runnables.
+        for (_, task) in active {
+            // Don't let a panicking waker blow everything up.
+            std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
```

Comment:
Why not? And how are we sure that this `AssertUnwindSafe` is indeed correct?

Reply:
I've figured out a way around the `AssertUnwindSafe`.

Reply:
There is still the initial question: why catch the panic here?

Reply:
Let's say a future panics. While the stack unwinds, the executor is dropped, which wakes up all of these wakers so that the tasks are all queued for the last time.

Since the future panicked, it's likely that something else was put into an invalid state and it's not out of the question that the waker will panic too. However, since the thread is currently panicking, a waker panic will be upgraded into an abort.

In the interest of not aborting, the waker panics here are just ignored.

Reply:
Could you add this explanation as a comment alongside this code then?